libvast/src/arrow_extension_types.cpp

```diff
@@ -0,0 +1,80 @@
+#include "vast/arrow_extension_types.hpp"
+
+#include <simdjson.h>
+
+namespace vast {
+
+namespace {
+
+auto deserialize(const std::string& json_) {
+  // simdjson requires additional padding on the string
+  simdjson::padded_string json{json_};
+
+  simdjson::dom::parser parser;
+  auto doc = parser.parse(json);
+  std::vector<struct vast::enumeration_type::field> enum_fields{};
+
+  for (auto f : doc.get_object()) {
+    std::string_view key = f.key;
+    enum_fields.emplace_back(std::string{key}, f.value.get_uint64());
+  }
+
+  return enumeration_type{enum_fields};
+}
+
+std::string serialize(const enumeration_type& t) {
+  auto out = fmt::memory_buffer();
+  auto inserter = std::back_inserter(out);
+  fmt::format_to(inserter, "{{ ");
+  bool first = true;
+  for (const auto& f : t.fields()) {
+    if (first)
+      first = false;
+    else
+      fmt::format_to(inserter, ", ");
+    fmt::format_to(inserter, "\"{}\": {}", f.name, f.key);
+  }
+  fmt::format_to(inserter, "}} ");
+
+  return fmt::to_string(out);
+}
+
+} // namespace
+
+std::string enum_extension_type::extension_name() const {
+  return "vast.enum";
+}
+
+std::shared_ptr<arrow::Array>
+enum_extension_type::MakeArray(std::shared_ptr<arrow::ArrayData> data) const {
+  VAST_ASSERT(data->type->id() == arrow::Type::EXTENSION);
+  VAST_ASSERT(ExtensionEquals(static_cast<const ExtensionType&>(*data->type)));
+  return std::make_shared<arrow::ExtensionArray>(data);
+}
+
+arrow::Result<std::shared_ptr<arrow::DataType>>
+enum_extension_type::Deserialize(std::shared_ptr<arrow::DataType> storage_type,
+                                 const std::string& serialized) const {
+  if (!storage_type->Equals(*storage_type_)) {
+    return arrow::Status::Invalid("Invalid storage type for VAST enum "
+                                  "dictionary: ",
+                                  storage_type->ToString());
+  }
+  return std::make_shared<enum_extension_type>(deserialize(serialized));
+}
+
+std::string enum_extension_type::Serialize() const {
+  return serialize(enum_type_);
+}
+
+void register_extension_types() {
+  if (const auto& t = make_arrow_enum(enumeration_type{{}});
```

Comment:
```suggestion
  if (const auto& t = make_arrow_enum(enumeration_type{{}});
```

Should this be allowed to fail, or do we need to `die` if it does?

Reply:
Modified to silently pass over data that's already been registered, but throws an error (`die`) when the extension has not been registered yet but registration fails.